book/src/internals/type-system.md

```diff
@@ -0,0 +1,323 @@
+
+  For now, this mapping is manual, but should be documented and may be
+  automated.
+
+- ... use any PRQL feature in their database. Here we are mapping back from PTS
+  into STS. Note that STS may have changed to a different dialect.
+
+  For example, translate PRQL's datetime operations to use TEXT in SQLite.
+
+  As of now, prql-compiler already does a good job of automatically doing this
+  mapping.
+
+Example of the mapping between PTS and two STSs:
+
+| PTS       | STS Postgres | STS SQLite |
+| --------- | ------------ | ---------- |
+| int32     | integer      | INTEGER    |
+| int64     | bigint       | INTEGER    |
+| timestamp | timestamp    | TEXT       |
+
+## Principles
+
+**Algebraic types** - have a way of expressing sum and product types. In Rust,
+sum would be an enum and product would be tuple or a struct. In SQL, product
+would be a row, since it can contain different types, all at once. Sum would be
+harder to express, see (this
+post)[https://www.parsonsmatt.org/2019/03/19/sum_types_in_sql.html ].
+
+The value proposition here is that algebraic types give a lot modeling
+flexibility, all while being conceptually simple.
+
+**Composable** - as with transformation, we'd want types to compose together.
+
+Using Python, JavaScript, C++ or Rust, one could define many different a data
+structure that would correspond to our idea of "relation". Most of them would be
+an object/struct that has column names and types and then a generic array of
+arrays for rows.
+
+PRQL's type system should also be able to express relations as composed from
+primitive types, but have only one idiomatic way of doing so.
+
+In practice this means that builtin types include only primitives (int, text,
+bool, float), tuple (for product), enum (for sum) and array (for repeating).
+
+An SQL row would translate to tuple, and a relation would translate to an array
+of tuples.
+
+I would also strive for the type system to be minimal - don't differentiate
+between tuples, objects and structs. Choose one and stick to it.
+
+**Type constraints** - constrain a type with a predicate. For example, have a
+type of int64 that are equal or greater than 10. Postgres
+[does support this](https://news.ycombinator.com/item?id=34835063). The primary
+value of using constrained types would not be validation (as it is used in
+linked article), but when matching the type.
+
+Say, for example, that we have a pipeline like this:
+
+```
+derive color = switch [x => 'red', true => 'green']
+derive is_red = switch [color == 'red' => true, color == 'green' => false]
+```
+
+It should be possible to infer that `color` is of type `text`, but only when
+equal to `'red'` or `'green'`. This means that the second switch covers all
+possible cases and `is_red` cannot be `null`.
+
+## Theory
+
+> For any undefined terms used in this section, refer to set theory and
+> mathematical definitions in general.
+
+A "type of a variable" is a "set of all possible values of that variable". This
```

Comment:
No strong opinion, but isn't the word `type` more familiar than the word `set`? I get the theory here. For the sake of our docs, I'd favor going back to `type` when these become docs....

Reply:
> going back to type

Changing to "or of two types represents a union of those two types"?

Reply:
My focus here iz the discussion below where we start to refer to types exclusively as sets. I support making the link between the concepts, I just think when we start documenting this, we should use the term "type" rather than "set".

Reply:
I see. I'm using term "set", because it's more suitable here in theory section and because it will be implemented this way - type declarations will be just variable declarations of sets.

Let's hold on with this until we convert this to documentation.